### PR TITLE
Add Electron Spell Check support

### LINKED ISSUE #2661: No spell check on the electron app version
Noticed this when I typed 'doesnt' which always highlights with a red line until you add the apostrophe. Once noticed, I was typing rubbish to test and make sure it wasn't a one off etc but nothing was highlighting.

### PR DESCRIPTION
**Pull request by MTRNord**
**DO NOT MERGE YET**

Add electron spellchecker. 
Features:

- [x] Underline wrong words
- [x] Change expected language based on choosen language or browser language
- [x] Support right click menu with options

This PR will fix #2661

https://github.com/atom/node-spellchecker/issues/72  blocks this feature for windows currently. It should in theory work on linux (not tested)

### CODE REVIEW

**Comment by MTRNord** [comment]
Oh well my branch wasn't clean... WiIl fix later

**Comment by MTRNord** [comment]
Test ran against master as I failed to use the correct base

**Comment by MTRNord** [comment]
This PR needs some User testing:

1. Try to use the new electron instructions told in the README.md (needs to be tested on multiple Systems (on linux requires gcc and X11))
2. Try to write bullshit, good text, and try if it works even after changing the language. (needs to be tested on multiple Systems)
3. Try the right click menu on not marked words in text fields and on marked works aswell as on textfields without selection.

File: electron_app/src/electron-main.js [hunk 1]
```diff
@@ -22,8 +22,11 @@ limitations under the License.
 const checkSquirrelHooks = require('./squirrelhooks');
 if (checkSquirrelHooks()) return;
 
+process.setMaxListeners(0);
```
Comment (t3chguy) [inline electron_app/src/electron-main.js]: why is this needed ooi?

File: electron_app/src/webcontents-handler.js [hunk 2]
```diff
@@ -84,14 +84,6 @@ function _CutCopyPasteSelectContextMenus(params) {
     }];
 }
 
-function onSelectedContextMenu(ev, params) {
```
Comment (t3chguy) [inline electron_app/src/webcontents-handler.js]: this cannot be right, you are removing the Selected context menu, how would highlighting some text need spellchecker?  Surely spellcheck is on inputs rather than highlighted text (which included non-inputs)

File: package.json [hunk 3]
```diff
@@ -79,6 +81,9 @@
     "ua-parser-js": "^0.7.10",
     "url": "^0.11.0"
   },
+  "peerDependencies": {
+    "electron-spellchecker": "^1.0.8"
```
Comment (t3chguy) [inline package.json]: little strange to have a dep as both a peerdep and a dep

File: src/vector/platform/ElectronPlatform.js [hunk 4]
```diff
@@ -129,6 +131,40 @@ export default class ElectronPlatform extends VectorBasePlatform {
...
+
+        async function showMenu(contextInfo) {
+            ipcRenderer.send('SelectedContextMenu', contextInfo);
+            ipcRenderer.once('SelectedContextMenu-reply', async function(event, arg) {
+                await generateMenu(contextInfo, arg);
+            })
+        }
+
+        let contextMenuListener = new ContextMenuListener(async function(info) {
+            await showMenu(info);
```
Comment (t3chguy) [inline src/vector/platform/ElectronPlatform.js]: not sure why this is an async function, its essentially just a normal fn which does `return showMenu(info)`

Comment (t3chguy) [inline src/vector/platform/ElectronPlatform.js]: could even be inlined to `(info) => showMenu(info)`

File: electron_app/src/webcontents-handler.js (hunk 2 above)
Comment (t3chguy) [inline electron_app/src/webcontents-handler.js]: my guess is the logic needs inverting and you need to override the input menu, but leave the highlighted text menu

**Comment by t3chguy** [comment]
Also another thing, this introduces a dependency to Python, just tripped me up on my python-less dev box, probably worth mentioning in the readme

File: package.json [hunk 5]
```diff
@@ -57,6 +58,7 @@
     "browser-request": "^0.3.3",
     "classnames": "^2.1.2",
     "draft-js": "^0.8.1",
+    "electron-rebuild": "^1.5.11",
```
Comment (t3chguy) [inline package.json]: is this not a dev dep?

**Comment by t3chguy** [comment]
Also looks like electron-spellchecker manually binds to document.body whereas electron has the underline stuff built into `electron.webFrame` so maybe worth looking into a lib that uses that such as https://github.com/superchu/electron-easy-spellcheck

File: electron_app/src/electron-main.js (hunk 1 above)
Comment (MTRNord) [inline electron_app/src/electron-main.js]: Because with spellchecker we have more listener than before. It starts to give overflows which causes to let things glitch

File: electron_app/src/webcontents-handler.js (hunk 2 above)
Comment (MTRNord) [inline electron_app/src/webcontents-handler.js]: THis moved to a place where I can use IPC as I need it 1. without spellcheck stuff and 2. with spellcheck stuff. As I cannot add the spellcheck options in Main I have to access it renderer. At this point I actually could fully move it to the renderer to keep stuff smaller.

File: package.json (hunk 3 above)
Comment (MTRNord) [inline package.json]: Will check

File: src/vector/platform/ElectronPlatform.js (hunk 4 above)
Comment (MTRNord) [inline src/vector/platform/ElectronPlatform.js]: It is so that I can use await. Without await it glitches the menu all over the place as the Spellcheck options are generated async

File: package.json (hunk 5 above)
Comment (MTRNord) [inline package.json]: Agreed

**Comment by MTRNord** [comment]
Didn't know of electron-easy-spellcheck I will check that out

File: electron_app/src/webcontents-handler.js (hunk 2 above)
Comment (t3chguy) [inline electron_app/src/webcontents-handler.js]: This is what you overrode too though: (Select, vs Input)
image

File: src/vector/platform/ElectronPlatform.js (hunk 4 above)
Comment (t3chguy) [inline src/vector/platform/ElectronPlatform.js]: `showMenu` is already an async function, I don't think both need to be
an async function returns a promise
so just returning `showMenu(...)` will be eq to `await showMenu(...)` in this situation

**Comment by ara4n** [comment]
@MTRNord it's unclear whether this is ready for review yet; please yell when it is.

**Comment by MTRNord** [comment]
@ara4n didn't yet have time to fix it yet. I will as soon as I got some time

**Comment by MTRNord** [comment]
To note: electron-easy-spellcheck isn't working at all. It doesn't seem to trigger the event it should trigger.

**Comment by MTRNord** [comment]
@t3chguy I updated this too changes I made localy a long time ago because of your comment. You might want to take a look at it :) PS: I have no Idea if it works at all because of the upstream bug and because of how old this code is already.

**Comment by MTRNord** [comment]
Also I need to change if this line is needed again or not. (as spellchecker is a native dep) `"electron:rebuild": "cd electron_app && electron-rebuild -v 1.6.8 -w electron-spellchecker && cd ..",`

**Comment by MTRNord** [comment]
We have to love nodejs...

```
Uncaught Error: Cannot open C:\Users\marce\WebstormProjects\riot-web\electron_app\node_modules\spellchecker\build\Release\spellchecker.node: Error: The module 'C:\Users\marce\WebstormProjects\riot-web\electron_app\node_modules\spellchecker\build\Release\spellchecker.node'
was compiled against a different Node.js version using
NODE_MODULE_VERSION 54. This version of Node.js requires
NODE_MODULE_VERSION 53. Please try re-compiling or re-installing
the module (for instance, using `npm rebuild` or`npm install`).
```

Basicly the electron version riot currently uses is compiled against a older versioin of nodejs

**Comment by MTRNord** [comment]
Current commit fails with `Unable to init SpellCheckerReferenceError: webFrame is not defined`

**Comment by MTRNord** [comment]
It works (no errors in console regarding spellchecking) But on windows 10 it does nothing because of the upstream bug. :( someone on linux might test. Keep in note to match the electron version installed and uses on rebuild

**Comment by MTRNord** [comment]
This test should have failed... It failed to compile

**Comment by MTRNord** [comment]
I am not able to maintain this PR further. Too less time and knowledge about changes in Riot since opening. Feel free to take the code for a new PR.
